chainstate/tx-verifier/src/transaction_verifier/mod.rs

```diff
@@ -582,6 +542,87 @@ impl<'a, S: TransactionVerifierStorageRef, U: UtxosView> TransactionVerifier<'a,
         }
     }
 
+    pub fn connect_pos_accounting_outputs(
+        &mut self,
+        tx_source: TransactionSource,
+        tx: &Transaction,
+    ) -> Result<(), ConnectTransactionError> {
+        let tx_undo = tx
+            .outputs()
+            .iter()
+            .filter_map(|output| match output.purpose() {
+                OutputPurpose::StakePool(data) => Some((data, output.value())),
+                OutputPurpose::Transfer(_)
+                | OutputPurpose::LockThenTransfer(_, _)
+                | OutputPurpose::Burn => None,
+            })
+            .map(
+                |(pool_data, output_value)| -> Result<PoSAccountingUndo, ConnectTransactionError> {
+                    let tx_id = tx.get_id();
+                    let total_inputs = self.calculate_total_inputs(tx.inputs())?;
+                    ensure!(
+                        total_inputs.contains_key(&CoinOrTokenId::Coin),
+                        ConnectTransactionError::MissingCoinOutputToStake
+                    );
+                    ensure!(
+                        total_inputs.len() == 1,
+                        ConnectTransactionError::TokenInputForPoSAccountingOperation(tx_id)
+                    );
+                    // TODO: check StakePoolData fields
+
+                    let input0 = tx.inputs().get(0).expect("must be some");
+                    let delegation_amount = output_value.coin_amount().ok_or(
+                        ConnectTransactionError::TokenOutputForPoSAccountingOperation(tx_id),
+                    )?;
+
+                    let (_, undo) = self
+                        .accounting_delta
+                        .create_pool(
+                            input0.outpoint(),
+                            delegation_amount,
+                            pool_data.decommission_key().clone(),
+                        )
+                        .map_err(ConnectTransactionError::PoSAccountingError)?;
+
+                    Ok(undo)
+                },
+            )
+            .collect::<Result<Vec<_>, _>>()?;
+
+        if !tx_undo.is_empty() {
+            self.accounting_delta_undo
+                .get_or_create_block_undo(&tx_source)
+                .insert_tx_undo(tx.get_id(), pos_accounting::AccountingTxUndo::new(tx_undo))
+                .map_err(ConnectTransactionError::AccountingBlockUndoError)
+        } else {
+            Ok(())
+        }
+    }
+
+    pub fn disconnect_pos_accounting_outputs(
```

Comment:
As discussed, please don't forget to test and verify that reorgs/disconnecting blocks restore the database state 100%.